File: Linked Lists/287. Find the Duplicate Number/287. Find the Duplicate Number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function (nums) {
  let fast = 0,
    slow = 0;

  while (true) {
    fast = nums[nums[fast]];
    slow = nums[slow];

    if (fast === slow) break;
  }

  slow = 0;
  while (fast !== slow) {
    fast = nums[fast];
    slow = nums[slow];
  }

  return slow;
};

// Time: O(n)
// Space: O(1)

File: Linked Lists/287. Find the Duplicate Number/287. Find the Duplicate Number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast = slow = 0

        while True:
            fast = nums[nums[fast]]
            slow = nums[slow]

            if fast == slow:
                break

        slow = 0
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]

        return slow

# Time: O(n)
# Space: O(1)
